README.md
# sanity-plugin-availability

Studio v3 exclusive Schema and Custom Input to indicate and rapidly author availability data.

![Custom input component showing days of the week and available times](https://raw.githubusercontent.com/SimeonGriggs/sanity-plugin-availability/main/img/availability.png)

## Installation

```
npm install --save sanity-plugin-availability
```

or

```
yarn add sanity-plugin-availability
```

## Usage

Add it as a plugin in sanity.config.ts (or .js):

```ts
// ./sanity.config.ts

 import {defineConfig} from 'sanity'
 import {availability} from 'sanity-plugin-availability'

 export const defineConfig({
  // all other settings...
  plugins: [
    // all other plugins...
    availability()
  ]
 })
```

Then use the `availability` field `type` as a field in your schema files.

```ts
defineField({
  name: 'availability',
  type: 'availability',
})
```

To create default times for all new documents, set an Initial Value at the document level. The helper function `defaultAvailabilityDays` is exported from the plugin for this.

(In future the plugin will do this itself, but a bug in the Studio presently requires this to be done at the document level so that the correct array `_key` values are set)

```ts
import {defineType} from 'sanity'
import {defaultAvailabilityDays} from 'sanity-plugin-availability'

export default defineType({
  name: 'person',
  title: 'Person',
  type: 'document',
  initialValue: {
    availability: defaultAvailabilityDays(),
  },
  // ... all other schema settings
})
```

## Caveats

To get this first version launched choices have been made and there are currently no configurable options:

- Default availability times are Monday-Friday, 9am-5pm.
- Time increments are only available in 15 minutes.
- Time is shown in 12 hour time, but can be searched by 24 hour time queries.
- Days must be sorted Monday to Sunday, times from earliest to latest, but there are no controls for adjusting sorting if they happen to get out of sync.

## License

MIT © Simeon Griggs
See LICENSE


## License

[MIT](LICENSE) © Simeon Griggs


## Develop & test

This plugin uses [@sanity/plugin-kit](https://github.com/sanity-io/plugin-kit)
with default configuration for build & watch scripts.

See [Testing a plugin in Sanity Studio](https://github.com/sanity-io/plugin-kit#testing-a-plugin-in-sanity-studio)
on how to run this plugin with hotreload in the studio.

### Release new version

Run ["CI & Release" workflow](https://github.com/SimeonGriggs/sanity-plugin-availability/actions/workflows/main.yml).
Make sure to select the main branch and check "Release new version".

Semantic release will only release on configured branches, so it is safe to run release on any branch.

CHANGELOG.md
<!-- markdownlint-disable --><!-- textlint-disable -->

# 📓 Changelog

All notable changes to this project will be documented in this file. See
[Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.0.1](https://github.com/SimeonGriggs/sanity-plugin-availability/compare/v1.0.0...v1.0.1) (2023-02-24)

### Bug Fixes

- remove quote from readme ([6a47b44](https://github.com/SimeonGriggs/sanity-plugin-availability/commit/6a47b443dec35eccea0b13d01b383344ab4d292a))
- remove schema warnings ([24abf81](https://github.com/SimeonGriggs/sanity-plugin-availability/commit/24abf8194f79bd5769cf3c63b70a44cdfbc49917))
- update image url in readme ([fff3b32](https://github.com/SimeonGriggs/sanity-plugin-availability/commit/fff3b325054bb39b54724be0e6bdcd7a80e119a9))

## 1.0.0 (2022-12-08)

### Features

- upgrade to plugin-kit ([b1472b8](https://github.com/SimeonGriggs/sanity-plugin-availability/commit/b1472b89d3247b15c6174eb90fa429d2343d42f3))

### Bug Fixes

- linting ([3fe9480](https://github.com/SimeonGriggs/sanity-plugin-availability/commit/3fe94807816a330e9a8eee6748155704c40dc584))


src/schema/availability.ts
import {defineField, Path} from 'sanity'

import Availability from '../components/Availability'
import {DAYS_OF_THE_WEEK} from '../helpers/data'
import {AvailabilityDay} from '../helpers/types'

export default defineField({
  name: 'availability',
  title: 'Availability',
  type: 'array',
  components: {
    // @ts-ignore
    input: Availability,
  },
  validation: (rule) =>
    rule
      .min(7)
      .max(7)
      .unique()
      .custom<AvailabilityDay[]>((value) => {
        if (!value?.length) {
          return true
        }

        const outOfOrderDays = value.reduce((acc, row, rowIndex) => {
          if (row.day === DAYS_OF_THE_WEEK[rowIndex]) {
            return acc
          }

          return [...acc, [{_key: row._key as string}] as Path]
        }, [] as Path[])

        return outOfOrderDays?.length
          ? {
              message: `Days must be in the correct order`,
              paths: outOfOrderDays,
            }
          : true
      }),
  of: [{type: 'availabilityDay', name: 'availabilityDay'}],
})


src/schema/availabilityDay.ts
import {defineField, Rule} from 'sanity'

import {HOURS_OF_THE_DAY, DAYS_OF_THE_WEEK} from '../helpers/data'
import {AvailabilityTime} from '../helpers/types'

export default defineField({
  name: 'availabilityDay',
  title: 'Day',
  type: 'object',
  fields: [
    {
      name: 'day',
      type: 'string',
      options: {
        list: DAYS_OF_THE_WEEK,
      },
    },
    {
      name: 'availableTimes',
      type: 'array',
      validation: (rule: Rule) =>
        rule.custom((times: AvailabilityTime[]) => {
          if (!times?.length) return true

          const timesOrdered = times.map(({from, to}) => {
            const fromIndex = HOURS_OF_THE_DAY.indexOf(from)
            const toIndex = HOURS_OF_THE_DAY.indexOf(to)

            return fromIndex + toIndex
          })
          const timesSorted = [...timesOrdered].sort((a, b) => a - b)

          return timesOrdered.join() === timesSorted.join()
            ? true
            : `Times must be sorted from earliest to latest`
        }),
      of: [{type: 'availabilityDuration', name: 'availabilityDuration'}],
    },
  ],
  preview: {
    select: {
      title: 'day',
      timesLength: 'availableTimes.length',
      times0from: 'availableTimes.0.from',
      times0to: 'availableTimes.0.to',
      times1from: 'availableTimes.1.from',
      times1to: 'availableTimes.1.to',
    },
    prepare(select) {
      const {title, timesLength, times0from, times0to, times1from, times1to} = select as Record<
        string,
        string
      >

      const timesDisplay = []
      if (times0from && times0to) {
        timesDisplay.push(`${times0from} – ${times0to}`)
      } else if (times1from && times1to) {
        timesDisplay.push(`${times1from} – ${times1to}`)
      } else if (!timesDisplay.length) {
        return {
          title,
          subtitle: `Unavailable`,
        }
      }
      return {
        title,
        subtitle:
          typeof timesLength === 'number' && timesLength > 2
            ? `${timesDisplay.join(', ')} and ${timesLength - 2} more`
            : timesDisplay.join(', '),
      }
    },
  },
})


src/schema/availabilityDuration.ts
import {defineField} from 'sanity'

import {HOURS_OF_THE_DAY} from '../helpers/data'
import {AvailabilityTime} from '../helpers/types'

export default defineField({
  name: 'availabilityDuration',
  title: 'Duration',
  type: 'object',
  options: {columns: 2},
  initialValue: {
    from: HOURS_OF_THE_DAY[36], // 9:00 AM
    to: HOURS_OF_THE_DAY[68], // 5:00 PM
  },
  validation: (rule) =>
    rule.custom((value) => {
      const {from, to} = (value as AvailabilityTime) ?? {}

      if (!from && !to) {
        return `Must contain both times`
      }

      const fromIndex = HOURS_OF_THE_DAY.indexOf(from)
      const toIndex = HOURS_OF_THE_DAY.indexOf(to)

      if (fromIndex === -1 || toIndex === -1) {
        return `Time must be selected from the list`
      }

      if (fromIndex >= toIndex) {
        return `"To" time must be later than "From" time`
      }

      return true
    }),
  fields: [
    {
      name: 'from',
      type: 'string',
      options: {
        list: HOURS_OF_THE_DAY,
      },
    },
    {
      name: 'to',
      type: 'string',
      options: {
        list: HOURS_OF_THE_DAY,
      },
    },
  ],
  preview: {
    select: {
      from: 'from',
      to: 'to',
    },
    prepare({from, to}) {
      return {
        title: from && to ? `${from} – ${to}` : `Incomplete time`,
      }
    },
  },
})


src/helpers/handleAutocompleteFilter.ts
type Option = {
  value: string
}

export const handleAutocompleteFilter = (query: string, option: Option): boolean => {
  // Return a search for `1.15` as `1:15`
  const queryFuzzy = query.includes(`.`) ? query.replace(`.`, `:`) : query
  const queryUpper = queryFuzzy.toUpperCase()

  // Return a search for `2a` as `2:00 AM`
  const shortTime = option.value
    .split(`:`)
    .map((t, i) => (i ? t.slice(3) : t))
    .join(``)

  // Return a search for `23:45` as `11:45 PM`
  const twentyFourHourTime = option.value.endsWith(`PM`)
    ? option.value
        .split(`:`)
        .map((t, i) => (i ? t : parseInt(t, 10) + 12))
        .join(`:`)
    : ``

  return (
    option.value.startsWith(queryUpper) ||
    twentyFourHourTime?.startsWith(queryUpper) ||
    shortTime.startsWith(queryUpper)
  )
}


src/helpers/data.ts
import {randomKey} from '@sanity/util/content'

import {AvailabilityDay, AvailabilityTime, Day, DayKey} from './types'

export const DAYS_OF_THE_WEEK: Day[] = [
  'Monday',
  'Tuesday',
  'Wednesday',
  'Thursday',
  'Friday',
  'Saturday',
  'Sunday',
]

export const dayKeysOfTheWeek: DayKey[] = [
  'monday',
  'tuesday',
  'wednesday',
  'thursday',
  'friday',
  'saturday',
  'sunday',
]

export const HOURS_OF_THE_DAY = [
  `12:00 AM`,
  `12:15 AM`,
  `12:30 AM`,
  `12:45 AM`,
  `1:00 AM`,
  `1:15 AM`,
  `1:30 AM`,
  `1:45 AM`,
  `2:00 AM`,
  `2:15 AM`,
  `2:30 AM`,
  `2:45 AM`,
  `3:00 AM`,
  `3:15 AM`,
  `3:30 AM`,
  `3:45 AM`,
  `4:00 AM`,
  `4:15 AM`,
  `4:30 AM`,
  `4:45 AM`,
  `5:00 AM`,
  `5:15 AM`,
  `5:30 AM`,
  `5:45 AM`,
  `6:00 AM`,
  `6:15 AM`,
  `6:30 AM`,
  `6:45 AM`,
  `7:00 AM`,
  `7:15 AM`,
  `7:30 AM`,
  `7:45 AM`,
  `8:00 AM`,
  `8:15 AM`,
  `8:30 AM`,
  `8:45 AM`,
  `9:00 AM`,
  `9:15 AM`,
  `9:30 AM`,
  `9:45 AM`,
  `10:00 AM`,
  `10:15 AM`,
  `10:30 AM`,
  `10:45 AM`,
  `11:00 AM`,
  `11:15 AM`,
  `11:30 AM`,
  `11:45 AM`,
  `12:00 PM`,
  `12:15 PM`,
  `12:30 PM`,
  `12:45 PM`,
  `1:00 PM`,
  `1:15 PM`,
  `1:30 PM`,
  `1:45 PM`,
  `2:00 PM`,
  `2:15 PM`,
  `2:30 PM`,
  `2:45 PM`,
  `3:00 PM`,
  `3:15 PM`,
  `3:30 PM`,
  `3:45 PM`,
  `4:00 PM`,
  `4:15 PM`,
  `4:30 PM`,
  `4:45 PM`,
  `5:00 PM`,
  `5:15 PM`,
  `5:30 PM`,
  `5:45 PM`,
  `6:00 PM`,
  `6:15 PM`,
  `6:30 PM`,
  `6:45 PM`,
  `7:00 PM`,
  `7:15 PM`,
  `7:30 PM`,
  `7:45 PM`,
  `8:00 PM`,
  `8:15 PM`,
  `8:30 PM`,
  `8:45 PM`,
  `9:00 PM`,
  `9:15 PM`,
  `9:30 PM`,
  `9:45 PM`,
  `10:00 PM`,
  `10:15 PM`,
  `10:30 PM`,
  `10:45 PM`,
  `11:00 PM`,
  `11:15 PM`,
  `11:30 PM`,
  `11:45 PM`,
]

export const defaultAvailabilityTimes = (): AvailabilityTime[] => [
  {
    _type: 'availabilityDuration',
    _key: randomKey(12),
    from: HOURS_OF_THE_DAY[36], // 9:00 AM
    to: HOURS_OF_THE_DAY[68], // 5:00 PM
  },
]

export const defaultAvailabilityDays = (): AvailabilityDay[] =>
  DAYS_OF_THE_WEEK.map((day, dayIndex) => ({
    _type: 'availabilityDay',
    _key: dayKeysOfTheWeek[dayIndex],
    day,
    availableTimes: [`Sunday`, `Saturday`].includes(day) ? [] : defaultAvailabilityTimes(),
  }))
